pkg/gcp/gcp.go

```diff
@@ -0,0 +1,156 @@
+func NewCloud(projectID, infraID string, client gcpclient.Interface) api.Cloud {
+	return &gcpCloud{
+		infraID:   infraID,
+		projectID: projectID,
+		client:    client,
+	}
+}
+
+// PrepareForSubmariner prepares submariner cluster environment on GCP
+func (gc *gcpCloud) PrepareForSubmariner(input api.PrepareForSubmarinerInput, reporter api.Reporter) error {
+	// create the inbound and outbound firewall rules for submariner public ports
+	reporter.Started("Opening public ports %q for cluster communications on GCP", formatPorts(input.PublicPorts))
+	ingress, egress := newExternalFirewallRules(gc.projectID, gc.infraID, input.PublicPorts)
+	if err := gc.openPorts(ingress, egress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened public ports %q with firewall rules %q and %q on GCP",
+		formatPorts(input.PublicPorts), ingress.Name, egress.Name)
+
+	// create the inbound firewall rule for submariner internal ports
+	reporter.Started("Opening internal ports %q for intra-cluster communications on GCP", formatPorts(input.PublicPorts))
+	internalIngress := newInternalFirewallRule(gc.projectID, gc.infraID, input.InternalPorts)
+	if err := gc.openPorts(internalIngress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened internal ports %q with firewall rule %q on GCP",
+		formatPorts(input.PublicPorts), internalIngress.Name)
+
+	return nil
+}
+
+// CleanupAfterSubmariner clean up submariner cluster environment on GCP
+func (gc *gcpCloud) CleanupAfterSubmariner(reporter api.Reporter) error {
+	ingressName, egressName := generateRuleNames(gc.infraID, publicPortsRuleName)
+
+	// delete the inbound and outbound firewall rules to close submariner public ports
+	reporter.Started("Deleting ingress firewall rule %q on GCP", ingressName)
+
+	if err := gc.client.DeleteFirewallRule(gc.projectID, ingressName); err != nil {
+		if !isGCPNotFoundError(err) {
+			reporter.Failed(err)
+			return err
+		}
+	}
+
+	reporter.Succeeded("Deleted ingress firewall rule %q on GCP", ingressName)
+
+	reporter.Started("Deleting egress firewall rule %q on GCP", egressName)
+
+	if err := gc.client.DeleteFirewallRule(gc.projectID, egressName); err != nil {
+		if !isGCPNotFoundError(err) {
+			reporter.Failed(err)
+			return err
+		}
+	}
+
+	reporter.Succeeded("Deleted egress firewall rule %q on GCP", egressName)
+
+	// delete the inbound and outbound firewall rules to close submariner internal ports
+	reporter.Started("Deleting ingress firewall rule %q on GCP", internalPortsRuleName)
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(gc.projectID, internalIngressName); err != nil {
+		if !isGCPNotFoundError(err) {
+			reporter.Failed(err)
+			return err
+		}
+	}
+
+	reporter.Succeeded("Deleted ingress firewall rule %q on GCP", internalPortsRuleName)
```

Comment:
This code is pretty repetitive with the other 2 calls which also delete public rules, it seems to me you can extract this to a function, eg

```go
func (gc *gcpCloud) deleteFirewallRule(name string) error {
	reporter.Started("Deleting firewall rule %q on GCP", name)
	if err := gc.client.DeleteFirewallRule(gc.projectID, name); err != nil {
		if !isGCPNotFoundError(err) {
			reporter.Failed(err)
			return err
		}
	}

	reporter.Succeeded("Deleted firewall rule %q on GCP", name)
}
```

You can then call it for each of the public ingress/egress and the private ingress

Reply:
good suggestion, updated